### PR TITLE
Add Azure Database Burst Simulation with Grafana integration in Demo project

### LINKED ISSUE #71: Implement Azure Database for PostgreSQL Simulation
# Implement Azure Database for PostgreSQL Simulation

## Summary
Implement simulation for Azure Database for PostgreSQL Flexible Server burstable instances (B-series), following the established AWS RDS simulation patterns. This enables modeling of CPU credit-based burst performance similar to AWS T-series instances, but adapted to Azure's baseline percentage model.

## Priority
**Medium** - New feature implementation

## Context

The AWS RDS simulation models T-series burstable instances with CPU credit mechanics. Azure Database for PostgreSQL Flexible Server offers similar B-series burstable instances with:
- **Baseline CPU percentage** (e.g., 20% for most B-series)
- **CPU credits** that enable bursting to 100% CPU
- **Credit earning/burning** mechanics similar to AWS

**Key insight:** Both AWS and Azure actually use the same baseline percentage model. The current AWS implementation incorrectly uses hardcoded `SlowSecs = 3.0s` (issue01.md #9), which makes it appear fundamentally different. In reality:
- AWS T3 medium: 20% baseline → `0.050 / 0.20 = 0.250s` (same as Azure!)
- Azure B2ms: 20% baseline → `0.050 / 0.20 = 0.250s`

This Azure implementation will use the correct baseline percentage model, which can later serve as a reference for fixing the AWS implementation.

## Scope

**Include:**
- B-series burstable instances (B1ms, B2s, B2ms, B4ms, B8ms)
- CPU credit earn/burn mechanics
- Baseline vs burst performance modeling
- CSV metrics export (latency, credits)
- CLI command for running simulations

**Exclude:**
- D-series (General Purpose) and E-series (Memory Optimized) - defer to future
- Unit tests - manual verification via CSV outputs
- AWS RDS fixes (tracked separately in issue01.md)

### 1. File Structure

Create new Azure-specific implementation mirroring AWS structure:

```
SimNextgenApp.Demo/AzurePgsqlSample/
├── AzurePgsqlInstanceSpec.cs    # Specs + registry
└── AzurePgsqlBehavior.cs        # Behavior with credit mechanics

SimNextgenApp.Demo/Scenarios/
└── AzurePgsqlBurstScenario.cs   # Scenario runner
```

Update: `SimNextgenApp.Demo/Program.cs` (add `azure-pgsql-burst` command)

### 2. Specification Classes

**File:** `SimNextgenApp.Demo/AzurePgsqlSample/AzurePgsqlInstanceSpec.cs`

```csharp
internal abstract record AzurePgsqlInstanceSpec(
    string Series,        // B, D, E
    string Size,          // 1ms, 2s, 2ms, 4ms
    int VCores,           // Virtual cores
    double BurstSecs,     // Service time at 100% CPU (exponential param)
    double BaselineSecs   // Service time at baseline CPU
);

internal record BurstableInstanceSpec(
    string Series,
    string Size,
    int VCores,
    double BurstSecs,
    double BaselineSecs,
    double BaselinePercent,      // e.g., 0.20 for 20% baseline
    double EarnRatePerHour,      // Credits earned per hour
    double MaxCredits            // Maximum credit balance
) : AzurePgsqlInstanceSpec(Series, Size, VCores, BurstSecs, BaselineSecs);

internal record FixedInstanceSpec(
    string Series,
    string Size,
    int VCores,
    double BurstSecs
) : AzurePgsqlInstanceSpec(Series, Size, VCores, BurstSecs, BurstSecs);

internal static class AzurePgsqlRegistry
{
    private static readonly List<AzurePgsqlInstanceSpec> Specs = new()
    {
        // B-series Burstable (20% baseline)
        new BurstableInstanceSpec("B", "1ms", 1, 0.050, 0.250, 0.20, 12, 288),
        new BurstableInstanceSpec("B", "2s", 2, 0.045, 0.225, 0.20, 24, 576),
        new BurstableInstanceSpec("B", "2ms", 2, 0.045, 0.225, 0.20, 24, 576),
        new BurstableInstanceSpec("B", "4ms", 4, 0.040, 0.200, 0.20, 48, 1152),
        new BurstableInstanceSpec("B", "8ms", 8, 0.035, 0.175, 0.20, 96, 2304),
    };

    public static AzurePgsqlInstanceSpec GetSpec(string series, string size)
    {
        var spec = Specs.FirstOrDefault(s => 
            s.Series.Equals(series, StringComparison.OrdinalIgnoreCase) && 
            s.Size.Equals(size, StringComparison.OrdinalIgnoreCase));
        
        if (spec == null)
            throw new ArgumentException($"Unknown Azure PostgreSQL instance: {series}.{size}");
        
        return spec;
    }
}
```

**Design Notes:**
- `BaselinePercent`: Explicit percentage (Azure model)
- `BaselineSecs = BurstSecs / BaselinePercent`: Service time at baseline performance
- `EarnRatePerHour` and `MaxCredits`: Azure-specific credit values

### 3. Behavior Class

**File:** `SimNextgenApp.Demo/AzurePgsqlSample/AzurePgsqlBehavior.cs`

**Improvements over current AWS implementation:**
- ✅ **Fix #1 (issue01.md):** Convert service time back to simulation units before returning
- ✅ **Fix #2 (issue01.md):** Remove redundant `timeUnit` parameter from `SetContext`
- ✅ **Fix #3 (issue01.md):** Initialize `_lastUpdateTime` from `context.ClockTime`
- ✅ **Fix #6 (issue01.md):** Remove unused `load` parameter from `GetServiceTime`
- ✅ **Fix #8 (issue01.md):** Throw exception instead of silent fallback in registry lookup
- ✅ **Fix #9 (issue01.md):** Use baseline percentage model instead of hardcoded SlowSecs

**Core Algorithm:**

```
1. Earn Credits:
   timeDelta = (context.ClockTime * multiplier) - lastUpdateTime  // in seconds
   earned = timeDelta * (EarnRatePerHour / 3600)
   credits = min(MaxCredits, credits + earned)
   lastUpdateTime = now

2. Look-Ahead Decision:
   estimatedCost = BurstSecs * (VCores / 60)
   canBurst = (credits >= estimatedCost)

3. Select Service Time:
   baseTime = canBurst ? BurstSecs : BaselineSecs
   actualDurationSecs = -baseTime * ln(1 - random)

4. Burn Credits:
   if canBurst:
     burnRate = VCores / 60  // per second
     actualBurn = actualDurationSecs * burnRate
     credits = max(0, credits - actualBurn)

5. Export Metrics:
   Log (timeInSeconds, latencyMs, credits) to CSV buffers

6. Return Service Time in Simulation Units:
   serviceTimeInSimUnits = actualDurationSecs / timeUnitToSecondsMultiplier
   return TimeSpan.FromTicks((long)(serviceTimeInSimUnits * TimeSpan.TicksPerSecond))
```

**Class Structure:**

```csharp
internal class AzurePgsqlBehavior
{
    private IRunContext? _engineContext;
    private double _credits;
    private double _lastUpdateTime = 0.0;
    private double _timeUnitToSecondsMultiplier = 1.0;
    
    private readonly AzurePgsqlInstanceSpec _spec;
    private readonly BurstableInstanceSpec? _burstableSpec;
    private readonly StringBuilder _latencyBuffer;
    private readonly StringBuilder _creditBuffer;
    
    public AzurePgsqlBehavior(AzurePgsqlInstanceSpec spec, double initialCredits = 10.0);
    public void SetContext(IRunContext context);      // Fixed: no timeUnit param
    public TimeSpan GetServiceTime(Random rnd);        // Fixed: no load param
    public void FinalizeExport(string directory);
    
    private double CalculateMultiplier(SimulationTimeUnit timeUnit);
    private void ExportMetrics(double nowInSeconds, double durationInSeconds);
}
```

### 4. Scenario Runner

**File:** `SimNextgenApp.Demo/Scenarios/AzurePgsqlBurstScenario.cs`

Follow AWS scenario structure with Azure-specific naming:

```csharp
internal static class AzurePgsqlBurstScenario
{
    public static void RunDemo(
        ILoggerFactory loggerFactory,
        double runDurationSecs,
        AzurePgsqlBehavior pgsqlBehavior,
        int genSeed)
    {
        // 1. Configure high-load generator (20 req/sec exponential)
        // 2. Configure server with pgsqlBehavior.GetServiceTime
        // 3. Create SimpleMmckModel (M/M/c/K queue)
        // 4. Setup SimulationProfile with TimeUnit validation
        // 5. Create SimulationEngine
        // 6. Call pgsqlBehavior.SetContext(engine)  // Fixed signature
        // 7. Run simulation
        // 8. Export metrics to output/ directory
    }
}
```

### 5. CLI Integration

**File:** `SimNextgenApp.Demo/Program.cs`

Add command after line 335 (after aws-rds-burst command):

```csharp
// ---- Demo: azure-pgsql-burst ----
var azurePgsqlBurstCommand = new Command("azure-pgsql-burst", 
    "Run Azure PostgreSQL Flexible Server burstable instance simulation");

var seriesOption = new Option<string>(
    name: "--series",
    description: "Azure instance series (B, D, E).",
    getDefaultValue: () => "B"
);

var azureSizeOption = new Option<string>(
    name: "--size",
    description: "Instance size (1ms, 2s, 2ms, 4ms, 8ms).",
    getDefaultValue: () => "2ms"
);

var azureDurationOption = new Option<double>(
    name: "--duration",
    description: "Total run duration in seconds.",
    getDefaultValue: () => 400.0
);

var azureInitialCreditsOption = new Option<double>(
    name: "--initial-credits",
    description: "Initial CPU credits for burstable instances.",
    getDefaultValue: () => 10.0
);

azurePgsqlBurstCommand.AddOption(seriesOption);
azurePgsqlBurstCommand.AddOption(azureSizeOption);
azurePgsqlBurstCommand.AddOption(azureDurationOption);
azurePgsqlBurstCommand.AddOption(azureInitialCreditsOption);

azurePgsqlBurstCommand.SetHandler((string series, string size, 
    double duration, double initialCredits) =>
{
    Log.Logger = new LoggerConfiguration()
        .MinimumLevel.Verbose()
        .Enrich.FromLogContext()
        .WriteTo.Seq("http://localhost:5341")
        .CreateLogger();

    loggerFactory = new LoggerFactory().AddSerilog(Log.Logger);
    
    var spec = AzurePgsqlRegistry.GetSpec(series, size);
    var pgsqlBehavior = new AzurePgsqlBehavior(spec, initialCredits);

    Console.WriteLine($"====== Running Azure PostgreSQL Burst Demo " +
        $"(Instance={series}.{size}, Duration={duration}s) ======");
    
    AzurePgsqlBurstScenario.RunDemo(loggerFactory, duration, 
        pgsqlBehavior, genSeed: 1234);
}, seriesOption, azureSizeOption, azureDurationOption, azureInitialCreditsOption);

// Add to demoCommand
demoCommand.AddCommand(azurePgsqlBurstCommand);
```

### Manual Testing Steps

1. **Build and run default scenario:**
   ```bash
   cd SimNextgenApp.Demo
   dotnet run demo azure-pgsql-burst --series B --size 2ms --duration 400 --initial-credits 10
   ```

2. **Expected outputs:**
   - Console shows simulation progress logs
   - `output/` directory created with CSV files:
     - `simulation_latency.csv` - timestamps and latency measurements
     - `simulation_credits.csv` - timestamps and credit balance

3. **Validate credit mechanics:**
   - **Initial phase:** Low latency (~45-50ms), credits depleting over time
   - **Depletion phase:** Credits reach zero, latency jumps to baseline (~225ms)
   - **Recovery phase:** If load decreases, credits regenerate, latency improves

4. **Test edge cases:**
   ```bash
   # Low initial credits (starts throttled)
   dotnet run demo azure-pgsql-burst --initial-credits 1
   
   # Different instance size
   dotnet run demo azure-pgsql-burst --size 4ms --initial-credits 20
   
   # Longer duration (observe full credit cycle)
   dotnet run demo azure-pgsql-burst --duration 800
   ```

5. **CSV analysis:**
   - Open `output/simulation_latency.csv` in Excel/Python/R
   - Plot latency over time → should show step change when credits deplete
   - Open `output/simulation_credits.csv`
   - Plot credits over time → should show linear decrease under load
   - Verify correlation: credits drop to zero → latency increases to baseline

6. **Compare with AWS RDS:**
   ```bash
   # Run equivalent AWS scenario
   dotnet run demo aws-rds-burst --family t3 --size medium --duration 400
   
   # Compare CSV outputs - patterns should be similar but with different values
   ```

## Acceptance Criteria

- [ ] Three new files created (spec, behavior, scenario)
- [ ] `azure-pgsql-burst` command added to Program.cs
- [ ] Command runs without errors
- [ ] CSV files generated in `output/` directory
- [ ] Credit depletion visible in `simulation_credits.csv`
- [ ] Latency increase visible in `simulation_latency.csv` when credits depleted
- [ ] All five AWS RDS fixes applied to Azure implementation
- [ ] Registry throws exception for unknown instance types (no silent fallback)

## Related Files

**New Files:**
- `SimNextgenApp.Demo/AzurePgsqlSample/AzurePgsqlInstanceSpec.cs`
- `SimNextgenApp.Demo/AzurePgsqlSample/AzurePgsqlBehavior.cs`
- `SimNextgenApp.Demo/Scenarios/AzurePgsqlBurstScenario.cs`

**Modified Files:**
- `SimNextgenApp.Demo/Program.cs`

**Reference Files:**
- `SimNextgenApp.Demo/AwsRdsSample/AwsRdsBehavior.cs` (pattern reference)
- `SimNextgenApp.Demo/AwsRdsSample/AwsRdsInstanceSpec.cs` (pattern reference)
- `SimNextgenApp.Demo/Scenarios/AwsRdsBurstScenario.cs` (pattern reference)
- `SimNextgenApp/Core/IRunContext.cs` (framework interface)
- `SimNextgenApp/Core/Utilities/SimulationTimeUnit.cs` (time unit conversion)

## Future Enhancements

- D-series (General Purpose) and E-series (Memory Optimized) support
- Unit tests for credit mechanics and time conversion
- Comparison reports between AWS and Azure performance
- Storage IOPS modeling (currently CPU-only)
- Automated visualization (PNG charts from CSV data)
- Fix AWS implementation to use baseline % model (making it directly comparable with Azure)
- Implement AWS unlimited mode with surplus credit tracking

### PR DESCRIPTION
**Pull request by goh-chunlin**

### CODE REVIEW

File: SimNextgenApp.Demo/AzureDbSample/AzureDbInstanceSpec.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+{
+    /// <summary>
+    /// Validates BaselineFraction and calculates SlowSecs safely.
+    /// </summary>
+    private static double ValidateAndCalculateSlowSecs(double fastSecs, double baselineFraction)
+    {
+        if (baselineFraction <= 0 || baselineFraction > 1.0)
+        {
+            throw new ArgumentOutOfRangeException(
+                "BaselineFraction",
```
[the remark this answers is not in the document]
Comment (goh-chunlin) [reply SimNextgenApp.Demo/AzureDbSample/AzureDbInstanceSpec.cs:38]: Will not fix

File: SimNextgenApp.Demo/AzureDbSample/AzureDbBehavior.cs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+        // Create histogram for latency distribution
+        _latencyHistogram = _meter.CreateHistogram<double>(
+            "sna.azure_db.query_latency",
+            unit: "ms",
+            description: "Query latency distribution in milliseconds");
+    }
+
+    public TimeSpan GetServiceTime(MyLoad load, Random rnd)
+    {
+        if (_engineContext == null) throw new InvalidOperationException("Context missing");
```
[the remark this answers is not in the document]
Comment (goh-chunlin) [reply SimNextgenApp.Demo/AzureDbSample/AzureDbBehavior.cs:83]: KIV.

✅ Not a bug - the code works correctly
✅ Only affects developers who misuse the API (rare)
✅ Error still thrown, just less helpful message
